### PR TITLE
DrivAerML: 4L/384d + lr=3e-4 — lower LR for wider model

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current best uses lr=5e-4, tuned on the original 256d model. When model capacity increases (256d→384d), the optimal learning rate often shifts lower — wider models have larger gradient norms per parameter and can benefit from more conservative updates. We also observed late-epoch oscillation in PR #2602 (epoch 141=10.2% → epoch 144=5.73%), which suggests the current LR may be too large for fine-grained convergence in the 384d regime. lr=3e-4 is the canonical Adam LR from many vision transformer papers; this tests whether a more conservative rate improves the final floor.

## Instructions

Run the following command exactly:

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 3e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent edward \
  --wandb-name "edward/drivaerml-4L384d-lr3e4" \
  --wandb-group "drivaerml-4L384d-variants"
```

Report the best `val_primary/surface_rel_l2_pct` and the epoch at which it occurred. If you observe reduced late-epoch oscillation compared to PR #2602, please note this explicitly — it will help us understand whether LR is driving the instability.

## Baseline

Current best metrics (PR #2602, W&B run `7ogfs7ph`):
- **val_primary/surface_rel_l2_pct: 5.73%** (epoch 144, 151 epochs total)
- Architecture: 4L/384d/6H, Fourier features enabled, no EMA, lr=5e-4
- Training note: late-epoch oscillation (epoch 141=10.2%, epoch 144=5.73%)
- Reproduce: `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 384 --model-heads 6 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`
- External target: **<3.71%** (AB-UPT benchmark)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### DrivAerML 4L/384d lr=3e-4 + T_max=30

| Metric | Baseline (lr=5e-4, PR #2602) | lr=3e-4 | Delta |
|--------|------------------------------|---------|-------|
| **val_primary/surface_rel_l2_pct** | **5.73%** | **12.15%** | +112% (worse) |
| test_primary/surface_rel_l2_pct | — | 12.95% | — |

- **Best val_primary/surface_rel_l2_pct:** 12.15% at **epoch 45** (cosine trough, LR=0)
- **Second best:** 13.35% at epoch 42
- **Epochs:** 45 (hit SENPAI_MAX_EPOCHS cap — still converging at cutoff)
- **W&B run:** je485sik
- **Training stability:** Notably smoother than lr=5e-4. Peak-to-trough oscillation typically 1-3%, with one spike to 8.3% at epoch 32. Compare to lr=5e-4 baseline which had 10.2%→5.73% late-epoch swings.

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 3e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 384 --model-heads 6 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent edward --wandb-name "edward/drivaerml-4L384d-lr3e4" \
  --wandb-group "drivaerml-4L384d-variants"
```

### What happened

**lr=3e-4 did NOT beat the baseline.** At 12.15% (epoch 45), it's more than 2x worse than the lr=5e-4 baseline of 5.73% (epoch 144). However, important context:

1. **Much fewer epochs:** This run only completed 45 epochs (hit epoch cap) vs the baseline's 151 epochs. At epoch 45, this run is still strongly converging — the trough envelope was 30.5→24.7→22.3→19.2→18.2→15.6→15.2→13.4→12.2% with no sign of plateauing.

2. **Smoother training:** The hypothesis about reduced oscillation is **confirmed**. lr=3e-4 shows peak-to-trough amplitudes of 1-3% (one outlier at 8.3%), compared to lr=5e-4's violent late-epoch swings (10.2%→5.73%). The lower LR produces a much more predictable convergence trajectory.

3. **Slower convergence per epoch:** The tradeoff is clear — lr=3e-4 converges more smoothly but more slowly. Whether it catches up or surpasses lr=5e-4 given the same number of epochs is unknown. At epoch 45, lr=5e-4 was likely in the ~12-13% range too (PR #2602 was still at 11.97% with the 256d model at epoch 34).

### Trough envelope (cosine minima)
| Epoch | val_primary/surface_rel_l2_pct |
|-------|-------------------------------|
| 6     | 35.74%                        |
| 8     | 30.52%                        |
| 13    | 24.66%                        |
| 15    | 23.08%                        |
| 17    | 22.35%                        |
| 20    | 20.02%                        |
| 22    | 19.19%                        |
| 24    | 18.18%                        |
| 29    | 15.61%                        |
| 31    | 15.18%                        |
| 38    | 13.95%                        |
| 40    | 13.39%                        |
| 42    | 13.35%                        |
| 45    | **12.15%**                    |

### Suggested follow-ups

1. **Longer training with lr=3e-4** — Give it 150+ epochs (matching the baseline) to see if the smoother trajectory catches up or surpasses lr=5e-4's best of 5.73%. The consistent ~1% improvement per trough cycle suggests significant headroom.
2. **lr=4e-4** — Split the difference between the stable 3e-4 and the volatile 5e-4. Might combine faster convergence with reduced oscillation.
3. **lr=3e-4 + 600 batches/epoch** — The current baseline improvement from 394→600 batches (PR #2645) was the biggest recent win. Combining more data per epoch with the smoother LR could be synergistic.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2693 because ADVISOR: Closing — 12.15% at 45ep (30-min budget) vs 5.73% at 151ep (180-min baseline). Budget mismatch makes this uninterpretable, and lr=3e-4 trajectory shows no advantage. edward, please pick up your next assignment..
